flink-core/src/main/java/org/apache/flink/api/common/operators/ResourceSpec.java

```diff
@@ -149,6 +150,39 @@ public ResourceSpec merge(final ResourceSpec other) {
 		return target;
 	}
 
+	/**
+	 * Subtracts another resource spec from this one.
+	 *
+	 * @param other The other resource spec to subtract.
+	 * @return The subtracted resource spec.
+	 */
+	public ResourceSpec subtract(final ResourceSpec other) {
+		checkNotNull(other, "Cannot subtract null resources");
+
+		if (this.equals(UNKNOWN) || other.equals(UNKNOWN)) {
+			return UNKNOWN;
+		}
+
+		checkArgument(other.lessThanOrEqual(this), "Cannot subtract a larger ResourceSpec from this one.");
+
+		final ResourceSpec target = new ResourceSpec(
+			this.cpuCores.subtract(other.cpuCores),
+			this.taskHeapMemory.subtract(other.taskHeapMemory),
+			this.taskOffHeapMemory.subtract(other.taskOffHeapMemory),
+			this.onHeapManagedMemory.subtract(other.onHeapManagedMemory),
+			this.offHeapManagedMemory.subtract(other.offHeapManagedMemory));
+
+		target.extendedResources.putAll(extendedResources);
+
+		for (Resource resource : other.extendedResources.values()) {
+			target.extendedResources.merge(resource.getName(), resource, (v1, v2) -> {
+				final Resource subtracted = v1.subtract(v2);
+				return subtracted.getValue().compareTo(BigDecimal.ZERO) == 0 ? null : subtracted;
```

Comment:
why are we using null instead of 0? `Resource` imposes no restriction on the value, hence 0 being a possible one. It is odd that , given 2 resources with a value of 0, the result of a subtraction is suddenly null.

Reply:
It's a good question. And I need to consider how we should treat a value 0 Resource.

CPUResource with value 0 is used in production currently, thus Resource must accept value 0. And CPUResource would not be null in ResourceSpec/ResourceProfile (except for the special UNKNOWN case where null means unknown value).

However, 0 value extends Resource is a bit confusing, since 0 and null also mean no such resource. If accepting both of them, we need to consider it in all comparisons(equals/lessThanOrEqual/isMatch) and aggregations(merge/subtract).

To be simple, I think we can change it like this: in `ResourceSpec`/`ResourceProfile` we only keep extended resource with positive values. If it is 0 value, we need to drop it so the resource would always be null. This change is needed in the constructors and `subtract()`.

Any usages of extended resources, must check whether it's null to see whether that extended resource is needed(for task)/available(in slot).

WDYT?

Reply:
This issue is not critical though, since extended resources are actually not supported in production yet. 
`TaskManagerServices#computeSlotResourceProfile` currently only creates slots with no extended resource, so a job with extended resources specified would not able to acquire slots to run tasks.

Reply:
I agree with forcing null if no resource is needed or available.

Reply:
FLINK-14869 is opened to track this issue.